cmd/portforward_tcp.go

```diff
@@ -96,6 +96,69 @@ func (p *TCPPortForwarder) Run(
 		}
 	}
 }
+func (p *TCPPortForwarder) handleInboundMessages(
+	ctx context.Context,
+	conn net.Conn,
+	sessionID string,
+	connectionID string,
+	ackChan <-chan struct{},
+	msgChan chan<- *ws.ProtoMsg,
+	recvChan <-chan *ws.ProtoMsg) {
+	sendStopMessage := true
+	defer func() {
+		conn.Close()
+		if sendStopMessage {
+			m := &ws.ProtoMsg{
+				Header: ws.ProtoHdr{
+					Proto:     ws.ProtoTypePortForward,
+					MsgType:   wspf.MessageTypePortForwardStop,
+					SessionID: sessionID,
+					Properties: map[string]interface{}{
+						wspf.PropertyConnectionID: connectionID,
+					},
+				},
+			}
+			msgChan <- m
+		}
+	}()
+
+	for {
+		select {
+		case m := <-recvChan:
+			if m.Header.Proto == ws.ProtoTypePortForward &&
+				m.Header.MsgType == wspf.MessageTypePortForwardStop {
+				sendStopMessage = false
+				return
+			} else if m.Header.Proto == ws.ProtoTypePortForward &&
+				m.Header.MsgType == wspf.MessageTypePortForward {
+				_, err := conn.Write(m.Body)
+				if err != nil {
+					if errors.Unwrap(err) != net.ErrClosed {
+						fmt.Fprintf(os.Stderr, "error: %v\n", err.Error())
+					}
+				} else {
+					// send the ack
+					m := &ws.ProtoMsg{
```

Comment:
this one shadows the m from the `case m := <-recvChan:` above, and also do we have to create it here every time? it is the same struct each time.